make it even more asynchronous

Risca opened the pull request:
This change makes afatfs_poll() an internal static function, and
everything else works (almost) as before. The one difference is that
callbacks may be called directly from the routine where it was provided,
even before the routine returns. E.g. the callback for "unlink complete"
may be called before afatfs_funlink() returns.

thenickdude commented on lib/asyncfatfs.c:
```diff
@@ -3407,9 +3440,19 @@ static void afatfs_initContinue()
             afatfs_createFile(&afatfs.freeFile, AFATFS_FREESPACE_FILENAME, FAT_FILE_ATTRIBUTE_SYSTEM | FAT_FILE_ATTRIBUTE_READ_ONLY,
                 AFATFS_FILE_MODE_CREATE | AFATFS_FILE_MODE_RETAIN_DIRECTORY, afatfs_freeFileCreated);
+            if (afatfs.filesystemState != AFATFS_FILESYSTEM_STATE_FATAL
+                && afatfs.initPhase != AFATFS_INITIALIZATION_FREEFILE_CREATING
```
This check against initPhase seems to be always false since initPhase is always AFATFS_INITIALIZATION_FREEFILE_CREATING here?

Risca replied:
`afatfs.initPhase` will be updated by the `afatfs_freeFileCreated()`, which now can happen before `afatfs_createFile()` returns

thenickdude commented:
My concern with this approach is that if the sdcard read/write primitives are implemented synchronously (i.e. they call the completion callback before returning), this will cause asyncfatfs to re-enter afatfs_poll(). This will lead to it performing an unbounded amount of work (particularly at card init when writing the freefile) and using an unbounded amount of stacksize, neither of which can be tolerated.

Synchronous read/write SD card primitives are a common choice when DMA is not available.

Risca commented:
> My concern with this approach is that if the sdcard read/write primitives are implemented synchronously (i.e. they call the completion callback before returning), this will cause asyncfatfs to re-enter afatfs_poll(). This will lead to it performing an unbounded amount of work (particularly at card init when writing the freefile) and using an unbounded amount of stacksize, neither of which can be tolerated.
> 
> Synchronous read/write SD card primitives are a common choice when DMA is not available.

Yes, I understand your concern. I had some thoughts about putting `ifdef`s around these changes since they are rather experimental, and some major disclaimers

Risca commented:
I reworked the changes so the feature can be enabled at compile time, and also updated the readme file :smiley_cat:

@thenickdude, do you think this feature is useful for AsyncFS?
